File: Python/observer.py
from abc import ABC, abstractmethod

class Observer(ABC):
    @abstractmethod
    def update(self, message: str):
        pass

class User(Observer):
    def __init__(self, name):
        self.name = name

    def update(self, message: str):
        print(f'{self.name} received message: {message}')

class NotificationChannel:
    def __init__(self):
        self.subscribers = []

    def subscribe(self, subscriber: Observer):
        self.subscribers.append(subscriber)

    def unsubscribe(self, subscriber: Observer):
        self.subscribers.remove(subscriber)

    def notify(self, message: str):
        for subscriber in self.subscribers:
            subscriber.update(message)


if __name__ == "__main__":
    # Użycie
    channel = NotificationChannel()

    alice = User('Alice')
    bob = User('Bob')

    channel.subscribe(alice)
    channel.subscribe(bob)

    channel.notify('Hello, World!')  # Alice and Bob receive notification

File: Cpp/observer.cpp
#include <iostream>
#include <vector>
#include <algorithm>

class Observer {
public:
    virtual ~Observer() {}
    virtual void update(const std::string& message) = 0;  // Pure virtual function
};

class User : public Observer {
public:
    User(std::string name) : name(name) {}
    void update(const std::string& message) override {
        std::cout << name << " received message: " << message << std::endl;
    }
private:
    std::string name;
};

class NotificationChannel {
public:
    void subscribe(Observer* observer) {
        subscribers.push_back(observer);
    }

    void unsubscribe(Observer* observer) {
        subscribers.erase(std::remove(subscribers.begin(), subscribers.end(), observer), subscribers.end());
    }

    void notify(const std::string& message) {
        for (Observer* subscriber : subscribers) {
            subscriber->update(message);
        }
    }
private:
    std::vector<Observer*> subscribers;
};

int main() {
    NotificationChannel channel;

    User alice("Alice");
    User bob("Bob");

    channel.subscribe(&alice);
    channel.subscribe(&bob);

    channel.notify("Hello, World!");  // Alice and Bob receive notification

    return 0;
}
